oeps/oep-0050-hooks-extension-framework.rst

```diff
@@ -0,0 +1,235 @@
+
+
+The wrapper should present functions to register the trigger call, to call the
+configured actions as a pipeline, and to defer calling the functions as async
+tasks when configured. This wrapper should make the triggers, especially those
+that do not allow for application flow change, to be compatible with the Async
+Event Messaging described in OEP-41.
+In terms of error handling, it should allow the configuration of silent failures
+letting operators know the source of the error.
+
+
+Rationale
+=========
+
+Hooks are a way for one piece of code to interact/modify another piece of code
+at specific, pre-defined spots. Many of the business rules of an installation
+can be altered and extended if the right spots are allowed to be extended by
+open edX plugins. Initially, this proposal is meant as a way for community
+members to extend the platform, but it could even become a way for the
+installation at edx.org to implement its particular business rules in a way that
+is sufficiently separated from the Open edX core.
+
+This proposal draws heavy inspiration from WordPress, in which the same
+`two types of hooks`_ are defined. Leveraging the Python/Django based technology
+of open edX makes actions a great candidate to be executed asynchronously in a
+different process as they are meant to be called with some context data, act on
+it, and return silently to the trigger location.
+Filters on the other hand are designed so that they are given some context data,
+act on it, and return a modified version for the rest of the core-code to use in
+the next steps of execution. We intend to make the two kinds of hooks
+interchangeable such that developers are not limited by the framework but
+instead given maximum flexibility on its usage to attain their extension goals.
+
+Hooks are used within WordPress to support almost all the forms in which it can
+be extended and are the basis of the very successful marketplace for plugins and
+themes. Given that Open edX already has a variety of theming capabilities and
+options we have opted here not to include any hooks that are meant to be used
+for theming.
+
+This proposal does not copy the model verbatim, since the open edX platform uses
+a different base technology which already allows for many extension points.
+This proposal deals primarily with the specific spots in the code execution that
+for many initiatives have been identified as the critical points that would
+allow them to adopt this technology without requiring modifications to the core.
+
+On the technical side, Django and the support for auto-installed plugins using
+stevedore already allow the addition of code in a very clean way. This proposal
+builds on top of this, only trying to standardize and grow the current
+capabilities of the platform.
+
+.. _two types of hooks: https://developer.wordpress.org/plugins/hooks/#actions-vs-filters
+
+
+Performance Considerations
```

Comment:
[request] Can we provide some basis for debugging and troubleshooting performance and functional errors? For example, can we include how one might support distributed tracing so we can track (1) how long each hook's implementation takes and (2) reliable traces/logs through multiple hooks.

Reply:
Added a paragraph at the end of this section about that.